A83-xds-gcp-authn-filter.md

```diff
@@ -0,0 +1,330 @@
+HTTP request completes.  When the HTTP request completes, the result
+(either success or failure, as described below) will be applied to all
+queued data plane RPCs.
+
+Note that when the token's expiration time is less than the refresh
+interval in the future, a new data plane RPC being started will trigger
+a new HTTP request, but the cached token value will still be used for
+that data plane RPC.  This pre-emptive re-fetching is intended to avoid
+periodic latency spikes when refreshing the token.
+
+If the HTTP request fails, all queued data plane RPCs will be failed
+with a gRPC status determined based on the returned HTTP status.  If the
+returned HTTP status maps to `UNAVAILABLE` in [HTTP to gRPC Status Code
+Mapping](https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md),
+then the data plane RPCs will be failed with status `UNAVAILABLE`;
+otherwise, they will be failed with status `UNAUTHENTICATED`.  If the
+request fails without an HTTP status (e.g., an I/O error), all queued
+data plane RPCs will be failed with `UNAVAILABLE` status.
+
+If the HTTP request succeeds, the body of the response will contain the
+JWT token. which the client will cache.  The client does not need to
+do full [RFC-7519] validation of the token (that is the responsibility
+of the server side), but it does need to extract the `exp` field for
+caching purposes.  If the `exp` field cannot be extracted (i.e., the JWT
+token is invalid), all queued data plane RPCs will be failed with status
+`UNAUTHENTICATED`.  Otherwise, the cache is updated, and the returned
+token is added to all queued data plane RPCs, which may then continue.
+
+If the HTTP request does not result in the cache being updated (i.e.,
+if the HTTP request fails or if it returns an invalid JWT token),
+[backoff](https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md)
+must be applied before the next attempt may be started.  If a data
+plane RPC is started when there is no cached token available and while
+in backoff delay, it will be queued until the completion of the next
+HTTP request attempt.  Note that no attempt should be started unless a
+data plane RPC is started, since we do not want to unnecessarily retry
+if the channel is idle.
+
+To add the token to a data plane RPC, the call credential will add a
+header named `authorization`.  The header value will be the string
+`Bearer ` (note trailing space) followed by the token value.
+
+### xDS HTTP Filter Configuration
+
+The xDS HTTP filter will be configured via the
+[`extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig`
+message](https://github.com/envoyproxy/envoy/blob/c16faca3619fb44c24b12d15aad8a797b9e210ab/api/envoy/extensions/filters/http/gcp_authn/v3/gcp_authn.proto#L27).
+The fields will be interpretted as follows:
+- `cache_config`: Optional.  Within this message:
+  - `cache_size`: Optional.  If set, must be greater than 0 and less
+    than `INT64_MAX`.  Defaults to 10.
```

Comment:
In other similar cases hasn't there been an upper limit to this sort of thing (even if implementation-defined)? Not that I think we need it for our security model or such, since this is xds-specific, but Envoy itself commonly had limits. I'm fine without, but it seemed strange/irregular.

Reply:
So this is a uint64 field that is limited to int64-1 range. Java will be limited to int32-1 range.

Is there really any benefit to NACKing too large of a value here? It seems silly; we should just clamp to our implementation limit. It seems the current limit is there because of an Envoy implementation limit.

Reply:
Oh, and Envoy explicitly supports `0` but we don't. There's a lot of strangeness here.

Reply:
Yeah, disabling the cache doesn't seem like a good idea to me (much less as a default!), since that could result in hammering the HTTP server.  It's also going to be more expensive memory-wise for us than for Envoy, since we're caching an entire call creds instance, not just the data for the token.  So I figured we'd use default 10 and not allow 0.

The upper bound was copied from Envoy; I didn't see any reason to change that.  I agree that it's odd that its implementation has that constraint, but I didn't have a good reason to do anything else in this case, so I opted for consistency.  But I agree that clamping to our max seems fine rather than NACKing in this case, and I've made that change.

Why would Java be limited to int32-1?  I know you don't have unsigned types, but you do have 64-bit types, right?  I would have expected you to have the same limitation as Envoy in that regard.

Reply:
Java is essentially 32 bit at its core. Arrays are limited to 31 bits of length. We'd need to use a two-tier list to handle larger values. The `-1` is for us to tell when we've exceeded the limit.

@shivaspeaks, note that this changed.

Reply:
Interesting, I didn't realize that.  In this case, the cache should be a map, not an array, but I don't know if that makes any difference.

I've changed the wording here to be a bit more generic.

Reply:
Hash maps are implemented using arrays. And the LRU is implemented using a hash map with a linked list between entries. Sure, trees are available, but we'd still need to do some custom work to have a 64 bit size and manage the LRU. I thought it was obvious there is no point in doing extra work to supporting things that size.

Reply:
Okay, so now it is generalized and it includes our case to cap the cache to int32-1. So I guess we don't need to change our LruCache? @ejona86 

> cache_size: Optional. If set, must be greater than 0. Defaults to 10. Implementations that cannot support caches as large as UINT64_MAX may cap this value at their maximum supported size.

Reply:
@shivaspeaks, let's discuss more in your PR, but the main thing is we no longer need to return-an-error/NAK if the value is too large; we'd only NAK on zero. Other cases we just clamp to our maximum.